File: ra-idp/adapters/persistence/redis/login-attempt-throttle.ts
/**
 * Layer 4 — Adapter Layer (Redis LoginAttemptThrottle)
 *
 * fixed-window counter + lockout キーの 2 段。詳細は ADR-029。
 *
 * Redis key:
 *   idp:login:counter:{kind}:{key}  (INCR + EXPIRE window_seconds, 初回のみ EXPIRE)
 *   idp:login:lock:{kind}:{key}     (SET NX EX lockout_seconds)
 *
 * PTTL で残り秒数を返すことで Retry-After を即時に組み立てる。
 * counter は window 内 INCR のみ。lockout 設定後は同じ key を削除して再カウント開始。
 */

import type {
  LoginAttemptThrottle,
  LoginThrottleAcquireResult,
  LoginThrottleKind,
} from '../../../src/authentication/ports/login-attempt-throttle'
import type { LoginThrottleConfig, LoginThrottleConfigs } from '../memory/login-attempt-throttle'
import type { Redis } from './client'

const COUNTER_PREFIX = 'idp:login:counter:'
const LOCK_PREFIX = 'idp:login:lock:'

export class RedisLoginAttemptThrottle implements LoginAttemptThrottle {
  constructor(
    private readonly redis: Redis,
    private readonly configs: LoginThrottleConfigs,
  ) {}

  async tryAcquire(
    kind: LoginThrottleKind,
    key: string,
    _now: Date,
  ): Promise<LoginThrottleAcquireResult> {
    const lockK = LOCK_PREFIX + kind + ':' + key
    const pttl = await this.redis.pttl(lockK)
    if (pttl > 0) {
      return { allowed: false, retryAfterSeconds: Math.ceil(pttl / 1000) }
    }
    return { allowed: true }
  }

  async recordFailure(
    kind: LoginThrottleKind,
    key: string,
    _now: Date,
  ): Promise<{ locked: boolean; retryAfterSeconds?: number }> {
    const config = this.configs[kind]
    const counterK = COUNTER_PREFIX + kind + ':' + key
    const lockK = LOCK_PREFIX + kind + ':' + key

    const failures = await this.redis.incr(counterK)
    if (failures === 1) {
      await this.redis.expire(counterK, config.windowSeconds)
    }
    if (failures >= config.maxFailures) {
      await this.redis.set(lockK, '1', 'EX', config.lockoutSeconds)
      await this.redis.del(counterK)
      return { locked: true, retryAfterSeconds: config.lockoutSeconds }
    }
    return { locked: false }
  }

  async recordSuccess(kind: LoginThrottleKind, key: string): Promise<void> {
    const counterK = COUNTER_PREFIX + kind + ':' + key
    const lockK = LOCK_PREFIX + kind + ':' + key
    await this.redis.del(counterK, lockK)
  }
}

export type { LoginThrottleConfig, LoginThrottleConfigs }

File: ra-idp-go/internal/infrastructure/persistence/memory/login_attempt_throttle.go
package memory

import (
	"context"
	"math"
	"sync"
	"time"

	authnports "ra-idp-go/internal/authentication/ports"
)

type LoginThrottleConfig struct {
	MaxFailures    int
	WindowSeconds  int
	LockoutSeconds int
}

type LoginThrottleConfigs struct {
	Account LoginThrottleConfig
	IP      LoginThrottleConfig
}

type loginCounter struct {
	failures  int
	expiresAt time.Time
}

type LoginAttemptThrottle struct {
	mu       sync.Mutex
	configs  LoginThrottleConfigs
	counters map[string]loginCounter
	locks    map[string]time.Time
}

func NewLoginAttemptThrottle(configs LoginThrottleConfigs) *LoginAttemptThrottle {
	return &LoginAttemptThrottle{
		configs: configs, counters: map[string]loginCounter{}, locks: map[string]time.Time{},
	}
}

func (t *LoginAttemptThrottle) TryAcquire(
	_ context.Context,
	kind authnports.LoginThrottleKind,
	key string,
	now time.Time,
) (authnports.LoginThrottleResult, error) {
	t.mu.Lock()
	defer t.mu.Unlock()
	lockKey := throttleKey(kind, key)
	expiresAt, ok := t.locks[lockKey]
	if !ok {
		return authnports.LoginThrottleResult{Allowed: true}, nil
	}
	remaining := expiresAt.Sub(now)
	if remaining <= 0 {
		delete(t.locks, lockKey)
		return authnports.LoginThrottleResult{Allowed: true}, nil
	}
	return authnports.LoginThrottleResult{
		Allowed: false, Locked: true,
		RetryAfterSeconds: int(math.Ceil(remaining.Seconds())),
	}, nil
}

func (t *LoginAttemptThrottle) RecordFailure(
	_ context.Context,
	kind authnports.LoginThrottleKind,
	key string,
	now time.Time,
) (authnports.LoginThrottleResult, error) {
	t.mu.Lock()
	defer t.mu.Unlock()
	config := t.config(kind)
	counterKey := throttleKey(kind, key)
	counter, ok := t.counters[counterKey]
	if !ok || !now.Before(counter.expiresAt) {
		counter = loginCounter{failures: 1, expiresAt: now.Add(time.Duration(config.WindowSeconds) * time.Second)}
	} else {
		counter.failures++
	}
	t.counters[counterKey] = counter
	if counter.failures < config.MaxFailures {
		return authnports.LoginThrottleResult{Allowed: true}, nil
	}
	delete(t.counters, counterKey)
	t.locks[counterKey] = now.Add(time.Duration(config.LockoutSeconds) * time.Second)
	return authnports.LoginThrottleResult{
		Allowed: false, Locked: true, RetryAfterSeconds: config.LockoutSeconds,
	}, nil
}

func (t *LoginAttemptThrottle) RecordSuccess(
	_ context.Context,
	kind authnports.LoginThrottleKind,
	key string,
) error {
	t.mu.Lock()
	defer t.mu.Unlock()
	key = throttleKey(kind, key)
	delete(t.counters, key)
	delete(t.locks, key)
	return nil
}

func (t *LoginAttemptThrottle) config(kind authnports.LoginThrottleKind) LoginThrottleConfig {
	if kind == authnports.LoginThrottleIP {
		return t.configs.IP
	}
	return t.configs.Account
}

func throttleKey(kind authnports.LoginThrottleKind, key string) string {
	return string(kind) + ":" + key
}
